### PR TITLE
✨ Attempt resize <amp-pinterest>

### PR DESCRIPTION
**Pull request by kristoferbaxter**
Addresses #10318.

This PR attempts to resize `<amp-pinterest>` on first layout completion for cases where the wrong height was specified.

### CODE REVIEW

File: extensions/amp-pinterest/0.1/pin-widget.js [hunk 1]
```diff
@@ -336,4 +342,17 @@ export class PinWidget {
...
+
+  /**
+   * Determine the height of the contents to allow resizing after first layout.
+   *
+   * @return {number|null}
+   */
+  height() {
+    return (
+      this.heightOwnerElement_
+        .getBoundingClientRect /* REVIEW */
```
Comment (dmanek) [inline extensions/amp-pinterest/0.1/pin-widget.js:354]: Does `/* REVIEW */` have any special meaning?

Comment (micajuine-ho) [inline extensions/amp-pinterest/0.1/pin-widget.js:354]: Should this height calculation be wrapped in a `vsync` mutator?

Comment (alanorozco) [reply extensions/amp-pinterest/0.1/pin-widget.js:354]: `/* REVIEW */` is a pointer for the reviewer to ensure that the use of a restricted API is correct.

When submitting, this should be changed to `/* OK */`. Without `/* OK */` the PR won't be able to be merged.

Comment (alanorozco) [reply extensions/amp-pinterest/0.1/pin-widget.js:354]: If possible I think we should not use `getBoundingClientRect` after all.

A single use InObs seems to be the best way to get this measurement without reflow/recalc.

```js
function getIntersectionEntryAsync(element) {
  return new Promise(resolve => {
    const inobs = new IntersectionObserver(entries => {
      inobs.disconect():
      resolve(entries[entries.length - 1]);
    });
    inobs.observe(element);
  });
}
```

I think @samouri introduced a util to the same effect recently.

Comment (samouri) [reply extensions/amp-pinterest/0.1/pin-widget.js:354]: Yep! It is called `measureIntersection`. 

https://github.com/ampproject/amphtml/blob/4da8b96882786565d2ae2d2725ed1c6ecb9fb300/src/utils/intersection.js#L58-L67

To get the height you'd call:
```js
measureIntersection(element).then(entry => entry.boundingClientRect.height);
```

---

Although TBH I'm not sure which is preferably right now. Using that or `measurePageLayoutBox` which does a measure in a vsync

https://github.com/ampproject/amphtml/blob/4da8b96882786565d2ae2d2725ed1c6ecb9fb300/src/utils/page-layout-box.js#L37-L44

File: extensions/amp-pinterest/0.1/follow-button.js [hunk 2]
```diff
@@ -90,4 +90,13 @@ export class FollowButton {
...
     return tryResolve(() => this.renderTemplate());
   }
+
+  /**
+   * Determine the height of the contents to allow resizing after first layout.
+   *
+   * @return {!Promise<number|null>}
+   */
+  height() {
+    return Promise.resolve().then(() => null);
```
Comment (alanorozco) [inline extensions/amp-pinterest/0.1/follow-button.js:100]: nit:
```suggestion
    return Promise.resolve(null);
```

File: extensions/amp-pinterest/0.1/save-button.js [hunk 3]
```diff
@@ -206,4 +206,13 @@ export class SaveButton {
...
     return promise.then(this.renderTemplate.bind(this));
   }
+
+  /**
+   * Determine the height of the contents to allow resizing after first layout.
+   *
+   * @return {!Promise<number|null>}
+   */
+  height() {
+    return Promise.resolve().then(() => null);
```
Comment (alanorozco) [inline extensions/amp-pinterest/0.1/save-button.js:216]:
```suggestion
    return Promise.resolve(null);
```

**Review by dmanek — APPROVED** [review]

**Pull request merged**